src/requirements/util/collections.ts

```diff
@@ -0,0 +1,87 @@
+export class HashMap<K, V> {
```

Comment:
I don't think I've ever had to use a HashMap in Javascript before - are you just using the Map() but setting the keys as hash values using the getUniqueHash function?

Reply:
Yes. I create a class for it to encapsulate the implementation detail so that the graph itself doesn't need to worry about it.

Reply:
Yup that makes sense!